### ISSUE TITLE
Multiplayer fails to start

### ISSUE THREAD

**Issue by LeNitrous** [ppy/osu#17477]
### Type

Game behaviour

### Bug description

Sometimes, the lobby fails to send players to the game. This is the relevant snippet of the logs that I have collected and appear from all players that participated from lobby of 40 players.

```
2022-03-27 09:12:32 [verbose]: ⚠️ One or more errors occurred. (An unexpected error occurred invoking 'ChangeState' on the server. InvalidStateChangeException: Cannot change from Idle to Loaded)
2022-03-27 09:12:32 [verbose]: Focus contention triggered by NotificationOverlay.
2022-03-27 09:14:18 [error]: An unobserved error has occurred.
2022-03-27 09:14:18 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (An unexpected error occurred invoking 'AbortGameplay' on the server. InvalidStateException: Cannot abort gameplay while not in a gameplay state)
2022-03-27 09:14:18 [error]: ---> Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'AbortGameplay' on the server. InvalidStateException: Cannot abort gameplay while not in a gameplay state
2022-03-27 09:14:18 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2022-03-27 09:14:18 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2022-03-27 09:14:18 [error]: --- End of inner exception stack trace ---
2022-03-27 09:14:18 [error]: An unobserved error has occurred. (inner)
2022-03-27 09:14:18 [error]: Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'AbortGameplay' on the server. InvalidStateException: Cannot abort gameplay while not in a gameplay state
2022-03-27 09:14:18 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2022-03-27 09:14:18 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2022-03-27 09:14:18 [verbose]: ⚠️ An unobserved error has occurred.
```
 

### Version

2022.327.0

### Logs

Full log:
[runtime.log](https://github.com/ppy/osu/files/8357263/runtime.log)

**Comment by bdach** [comment]
I've had this happen to me once on the release build but the circumstances are unclear to me as to why. Can't reproduce on dev server, for one thing.

**Comment by LeNitrous** [comment]
It happened very frequently on a large lobby to multiple people that I was in. I don't know what causes this but the side-effects are players unable to "unready" until they leave and rejoin the lobby.

**Comment by peppy** [comment]
Probably need to figure whether this is a new issue or one that has existed before the release yesterday.

**Comment by smoogipoo** [comment]
This should’ve been fixed via https://github.com/ppy/osu-server-spectator/pull/108…

**Comment by smoogipoo** [comment]
I found the reason why this is still happening:

`MultiplayerHub.changeAndBroadcastUserState()` uses the current connection ID to add/remove players from user groups: https://github.com/ppy/osu-server-spectator/blob/b18144df5da7c447051cf479ddc4cc02d489d9cc/osu.Server.Spectator/Hubs/MultiplayerHub.cs#L725-L736

In most cases this is fine, _except_ when `MultiplayerHub.unreadyAllUsers()` is called, which updates all users in the room and can't use only the current connection context. So the failure scenario here is:
1. Join player 1 (host), player 2 (guest).
2. Ready up player 2.
3. Change room settings on player 1 (e.g. the queue mode). Player 2 will unready.
4. Start play _without_ ready-ing up player 2.

I'm in the process of refactoring how external callbacks are done in the multiplayer hub, so this should be fixed as part of that.

**Comment by respektive** [comment]
Same thing still happens when the host is changing the beatmap while a guest has already ready up. The guest won't unready automatically and fail to enter gameplay.

https://user-images.githubusercontent.com/22381687/160741998-e200f688-3c02-4d03-964b-7eff5cb74d53.mp4

**Comment by peppy** [comment]
This change has not been deployed yet. It should happen today or tomorrow.

**Comment by respektive** [comment]
Alright that's my bad then, thought it has been since the issue has been closed. :smile:
